src/index/IndexImpl.cpp

```diff
@@ -1338,28 +1337,24 @@ const Permutation& IndexImpl::getPermutation(Permutation::Enum p) const {
 
 // __________________________________________________________________________
 Index::NumNormalAndInternal IndexImpl::numDistinctSubjects() const {
-  if (hasAllPermutations()) {
-    auto numActually = numSubjectsNormal_;
-    return {numActually, spo_.metaData().getNofDistinctC1() - numActually};
-  } else {
-    AD_THROW(
-        "Can only get # distinct subjects if all 6 permutations "
-        "have been registered on sever start (and index build time) "
-        "with the -a option.");
-  }
+  AD_CONTRACT_CHECK(
+      hasAllPermutations(),
+      "Can only get # distinct subjects if all 6 permutations "
+      "have been registered on sever start (and index build time) "
+      "with the -a option.");
```

Comment:
I understand that this is an AD_CONTRACT_CHECK in the sense that this function shouldn't have been called if the permutations necessary to compute it are not loaded.

But isn't this abused here for checking the validity of a user request?

For example, when there is a syntax error in the SPARQL query, we also don't report this via an AD_CONTRACT_CHECK (although we could with the justification that the contract of the respective function is that it should be called with a syntactically correct SPARQL query).

Reply:
The code that calls this should (and as far as I've just checked always does) check for `hasAllPermutations()` first. 
So the nature of this being a `CONTRACT_CHECK` seems legitimate to me.
I however did not really analyze this closely, but merely replaced some occurences of `if(....) {AD_THROW(...)}` with the new syntax.